lib/libp2p/libp2p.ex

```diff
@@ -186,35 +213,66 @@ defmodule Libp2p do
   Returns false if there are no more nodes.
   """
   @spec iterator_next(iterator) :: boolean
-  def iterator_next(_iterator),
-    do: :erlang.nif_error(:not_implemented)
+  def iterator_next(_iterator), do: :erlang.nif_error(:not_implemented)
 
   @doc """
   Returns the current node.
   WARN: you need to call iterator_next before calling this function!
   """
   @spec iterator_node(iterator) :: {:ok, discv5_node} | error
-  def iterator_node(_iterator),
-    do: :erlang.nif_error(:not_implemented)
+  def iterator_node(_iterator), do: :erlang.nif_error(:not_implemented)
 
   @doc """
   Returns the published TCP port of the node, or nil.
   """
   @spec node_tcp(discv5_node) :: integer | nil
-  def node_tcp(_node),
-    do: :erlang.nif_error(:not_implemented)
+  def node_tcp(_node), do: :erlang.nif_error(:not_implemented)
 
   @doc """
   Returns the multiaddresses of the node.
   """
   @spec node_multiaddr(discv5_node) :: {:ok, addrs} | error
-  def node_multiaddr(_node),
-    do: :erlang.nif_error(:not_implemented)
+  def node_multiaddr(_node), do: :erlang.nif_error(:not_implemented)
 
   @doc """
   Returns the ID of the node.
   """
   @spec node_id(discv5_node) :: {:ok, peer_id} | error
-  def node_id(_node),
-    do: :erlang.nif_error(:not_implemented)
+  def node_id(_node), do: :erlang.nif_error(:not_implemented)
+
+  @doc """
+  Creates a new GossipSub router.
+  """
+  @spec new_gossip_sub(host) :: {:ok, pub_sub} | error
+  def new_gossip_sub(_host), do: :erlang.nif_error(:not_implemented)
+
+  @doc """
+  Joins a topic. If the topic was already joined, this will fail.
+  """
+  @spec pub_sub_join(pub_sub, binary) :: {:ok, topic} | error
+  def pub_sub_join(_pubsub, _topic_str), do: :erlang.nif_error(:not_implemented)
+
+  @doc """
+  Subscribes to the given topic.
+  """
+  @spec topic_subscribe(topic) :: {:ok, subscription} | error
+  def topic_subscribe(_topic), do: :erlang.nif_error(:not_implemented)
+
+  @doc """
+  Publishes data on the given topic.
+  """
+  @spec topic_publish(topic, binary) :: :ok | error
+  def topic_publish(_topic, _data), do: :erlang.nif_error(:not_implemented)
+
+  @doc """
+  Gets next message of given subscription.
+  """
+  @spec subscription_next(subscription) :: {:ok, message} | error
+  def subscription_next(_subscription), do: :erlang.nif_error(:not_implemented)
+
+  @doc """
+  Gets the application data from a message.
```

Comment:
what is the application data from a message?

Reply:
It's the raw data sent by a peer, without any of the headers added by pubsub